File: electroplanetaudit/src/main/java/com/PFE/electroplanetaudit/service/StoreService.java
package com.PFE.electroplanetaudit.service;

import com.PFE.electroplanetaudit.entity.Store;
import com.PFE.electroplanetaudit.repository.StoreRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import java.time.LocalDateTime;
import java.util.List;

@Service
@RequiredArgsConstructor
@Transactional
public class StoreService {

    private final StoreRepository storeRepository;

    // Create new store (code is required from frontend)
    public Store create(Store store) {
        store.setDateCreation(LocalDateTime.now());
        store.setActif(true);
        return storeRepository.save(store);
    }

    // Get all stores (for export to BI)
    @Transactional(readOnly = true)
    public List<Store> findAll() {
        return storeRepository.findAll();
    }

    // Get store by ID
    @Transactional(readOnly = true)
    public Store findById(Long id) {
        return storeRepository.findById(id).orElse(null);
    }

    // Get store by code
    @Transactional(readOnly = true)
    public Store findByCode(String code) {
        return storeRepository.findByCode(code).orElse(null);
    }

    // Get active stores only
    @Transactional(readOnly = true)
    public List<Store> findActiveStores() {
        return storeRepository.findByActifTrue();
    }

    // Search by keyword (for search bar)
    @Transactional(readOnly = true)
    public List<Store> search(String keyword) {
        if (keyword == null || keyword.trim().isEmpty()) {
            return storeRepository.findAll();
        }
        return storeRepository.searchByKeyword(keyword);
    }

    // Get stores with filters and pagination (for store list page)
    @Transactional(readOnly = true)
    public Page<Store> getStoresWithFilters(String keyword, Boolean actif, Pageable pageable) {
        return storeRepository.findAllWithFilters(keyword, actif, pageable);
    }


    @Transactional(readOnly = true)
    public List<Store> findByRegion(String region) {
        return storeRepository.findByRegion(region);
    }

    // Get all regions (for filter dropdown)
    @Transactional(readOnly = true)
    public List<String> getAllRegions() {
        return storeRepository.findAllRegions();
    }

    // Get all villes (for filter dropdown)
    @Transactional(readOnly = true)
    public List<String> getAllVilles() {
        return storeRepository.findAllVilles();
    }

    // Get store statistics for BI
    @Transactional(readOnly = true)
    public List<Object[]> getStoresByRegion() {
        return storeRepository.countStoresByRegion();
    }

    // FULL UPDATE - admin can edit all fields
    public Store update(Long id, Store updatedStore) {
        Store existing = storeRepository.findById(id).orElse(null);
        if (existing == null) {
            return null;
        }

        // Update all fields
        if (updatedStore.getCode() != null) existing.setCode(updatedStore.getCode());
        if (updatedStore.getNom() != null) existing.setNom(updatedStore.getNom());
        if (updatedStore.getAdresse() != null) existing.setAdresse(updatedStore.getAdresse());
        if (updatedStore.getVille() != null) existing.setVille(updatedStore.getVille());
        if (updatedStore.getRegion() != null) existing.setRegion(updatedStore.getRegion());
        if (updatedStore.getActif() != null) existing.setActif(updatedStore.getActif());

        return storeRepository.save(existing);
    }

    // Delete store
    public boolean delete(Long id) {
        if (!storeRepository.existsById(id)) {
            return false;
        }
        storeRepository.deleteById(id);
        return true;
    }
}

File: my-app/services/storeService.js
import api from "@/lib/axios"

export const storeService = {
  // GET /api/stores/filtered?page=0&size=10&sortBy=nom&sortDir=asc&keyword=casa&actif=true
  getFiltered: (params) =>
    api.get("/api/stores/filtered", { params }),

  // GET /api/stores
  getAll: () =>
    api.get("/api/stores"),

  // GET /api/stores/:id
  getById: (id) =>
    api.get(`/api/stores/${id}`),

  // GET /api/stores/active
  getActive: () =>
    api.get("/api/stores/active"),

  // GET /api/stores/search?keyword=casa
  search: (keyword) =>
    api.get("/api/stores/search", { params: { keyword } }),

  // GET /api/stores/regions  → for dropdowns
  getRegions: () =>
    api.get("/api/stores/regions"),

  getByRegion: (region) =>
    api.get(`/api/stores/by-region/${region}`), 

  // GET /api/stores/villes  → for dropdowns
  getVilles: () =>
    api.get("/api/stores/villes"),

  // GET /api/stores/stats/regions  → BI
  getStatsByRegion: () =>
    api.get("/api/stores/stats/regions"),

  // POST /api/stores  (ADMIN only)
  create: (data) =>
    api.post("/api/stores", data),

  // PUT /api/stores/:id  (ADMIN only)
  update: (id, data) =>
    api.put(`/api/stores/${id}`, data),

  // DELETE /api/stores/:id  (ADMIN only)
  delete: (id) =>
    api.delete(`/api/stores/${id}`),
}